### PR TITLE
Collect data harvard

### PR DESCRIPTION
**Pull request by ygxu01**
I will refine the transform file (split the clean function into clean, standardize, and create_table etc.) and create the pipeline next week :)

### CODE REVIEW

**Review by sarahwalker10 — APPROVED** [review]
Message to Yangge: Great work this week, Yangge! One thing: can you please upload the Harvard data you used to the Drive, that way we can all access the data? Thank you! 

Review for Trevor: Yangge found an election results dataset and wrote a script to clean and standardize the data. She did a good job creating functions in the HarvardTransformer class (`src/utils/transform/harvard.py`) that have the same name and purpose as the other state transformer class functions; that way, the pipeline can be run on the Harvard dataset the same way as the other state dfs because it will be calling the same functions to clean and transform the data.

**Review by trevorspreadbury — CHANGES_REQUESTED** [review]
I'm guessing this is 90% there, but I can't merge this in right now because it doesn't meet minimum viable pull request standard. I can't run the notebook because it depends on a 'harvard.csv' file that is not documented. 

Also the harvard data code should not be part of the StateTransformer pipeline. Ping @sarahwalker10 and I when this is done and we'll merge it in!

**Review by sarahwalker10 — COMMENTED** [review]
Hi Yangge, Great work!! You've clearly put in a lot of effort not only on the StateTransformers, but also on creating a separate pipeline process for the election data. Here are my comments:
1) In the TA session on Friday I'd like to have you share your screen and run the harvard_eda.ipynb notebook + the transform_election_pipeline.py script you wrote. I am getting an error when I try to run these files, but it could be an issue on my end. Either way we can debug as a group! 
2) I'm curious if everything related to the election results should be put in its own directory (for ex, create a new path for src/utils/election_transform) and take these scripts out of the current path src/utils/transform (which is meant for campaign finance data). 
3) I can't open the .docx file that you uploaded for the data/codebook dictionary of the Harvard dataset. I recommend removing the codebook word doc file from GitHub and uploading that to the drive as well, alongside the dataset. Furthermore, please rename the files (both the datafile and the codebook file) to be more clear. 

Also, in the utils/transform/harvard.py script, you commented " # Shall I put this function in this file or in transform file?" on line 94 (in regards to the uuid). Looking at the code for the state transformers, it seems that the uuid assignment is handled within the state cleaning scripts (i.e. michigan.py, minnesota.py, etc.), before it reaches the pipeline.py file where transform_and_merge() is created. So to answer your question, I think you have the function in the right place. 

Great work! I know you are continuing to improve this week, so looking forward to seeing these updates at the TA session!
-Sarah

File: src/utils/transform/harvard.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        rows not representing minimal viable transactions
+
+        Inputs:
+            data: Dataframe as output from preprocess method.
+
+        Returns: Dataframe 
+        """
+        clean_df = data.copy(deep=True)
+        clean_df = clean_df[HV_INDIVIDUAL_COLS] 
+        clean_df = clean_df[clean_df["year"] <=2016 & clean_df["year"] >=2014]
```
Comment (trevorspreadbury) [inline src/utils/transform/harvard.py]: & is evaluated before <= so this is equal to `clean_df[clean_df["year"] <=(2016 & clean_df["year"]) >=2014]` which will lead to an error. Did this run for you?

File: src/utils/transform/harvard.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        data["last_name"] = data["last"].str.upper()
+        data["first_name"] = data["first"].str.upper()
+        data["full_name"] = data["cand"].str.lower()
+        data["full_name"] = data["full_name"].astype(str)[
+            data["full_name"].notna()
+        ]
+        data = data.apply(standardize_individual_names, axis = 1)
+
+        party_map = {"d": "democratic", "r": "republican", "partymiss": np.nan}
+        data["party"] = data["party"].map(party_map)
```
Comment (trevorspreadbury) [inline src/utils/transform/harvard.py]: theres' no column named party -- is this supposed to be partyt?

File: src/utils/transform/harvard.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        Returns:
+            a table as defined in database schema
+        """
+        final_table = data.copy()
+
+        final_table = self.create_election_result_uuid(final_table)
+
+        return final_table
+    
+    # Shall I put this function in this file or in transform file?
```
Comment (trevorspreadbury) [inline src/utils/transform/harvard.py]: this is fine here

**Review by trevorspreadbury — CHANGES_REQUESTED** [review]
There is a lot of good stuff in here. The documentation of the new dataset is good and the election pipeline is well thought out and a unique way of doing it. The standardize_individual_names looks like a good addition, and HumanName looks like a good find. 

I'll echo Sarah and say that the election transform pipeline should be in its own directory (package). 

The main problem is that this is not meeting a minimum viable PR right now. The linter is failing and the code runs into multiple errors when I try to run it. I've added comments that highlight these problems. Make sure you fix these for next week to avoid losing points and do reach out if you need help.

**Review by sarahwalker10 — COMMENTED** [review]
Great work this week, Yangge! And good incorporation of the feedback from last week. Here are my comments this week:
1. Your code is still not passing the pre-commit checks - do you have any questions about how to check for this or how to make the requested changes by the linter? 
2. A small correction needs to be made on line 8 of `utils/election/constants.py`: because you renamed the harvard data file, you'll need to update this for `HV_FILEPATH`. (Once this adjustment gets made, your code for `scripts/transform_election_pipeline.py` runs successfully, but does throw a warning about decoding)
3. Trying to run `scripts/transform_pipeline.py` and I get an error on the PA state transformer. Specifically:

 ```
File "/project/src/utils/transform/pennsylvania.py", line 76, in preprocess
    year = int(year_directory.stem)
           ^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'contrib'
```

4. Your new notebook `notebooks/election_dedupe.ipynb` imports a package called `altair` : I'm not sure where this is being used in your code, but if you do plan to use it moving forward, it will need to be added to the `requirements.txt` file!

Great job exploring the splink package by the way!

**Pull request merged**